source/Timer.cpp

```diff
@@ -0,0 +1,253 @@
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			out.Write("proximity", proximity, distance);
+			if(proximityCenter != nullptr)
+			{
+				out.BeginChild();
+				{
+					out.Write(proximityCenter->Name());
+				}
+				out.EndChild();
+			}
+		}
+
+		action.Save(out);
+	}
+	out.EndChild();
+}
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action
+Timer Timer::Instantiate(map<string, string> &subs,
+						const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.name = name;
+	result.base = base;
+	result.rand = rand;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+
+	result.timeToWait = timeToWait;
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	result.action = action.Instantiate(subs, origin, jumps, payload);
+
+	return result;
+}
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+void Timer::ResetOn(ResetCondition cond)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		timeToWait = base + Random::Int(rand);
+	}
+}
+
+void Timer::Step(PlayerInfo &player, UI *ui)
+{
+	if(isComplete)
+		return;
+	if((system && player.Flagship()->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(player.Flagship()->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM);
+		isActive = false;
+		return;
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!player.Flagship()->IsThrusting() && !player.Flagship()->IsSteering()
+						&& !player.Flagship()->IsReversing());
+		for(const Hardpoint &weapon : player.Flagship()->Weapons())
+			shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE);
+			isActive = false;
+			return;
+		}
+	}
+	if(requireUncloaked)
+	{
+		double cloak = player.Flagship()->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE);
+			isActive = false;
+			return;
+		}
+	}
+	if(proximity > 0.)
+	{
+		Point center;
+		if(proximityCenter)
+		{
+			const StellarObject *proximityObject = system->FindStellar(proximityCenter);
+			center = proximityObject->Position();
+		}
+		double dist = (player.Flagship()->Position() - center).Length();
+		if((closeTo && dist > proximity) || (!closeTo && dist < proximity))
+		{
+			ResetOn(Timer::ResetCondition::LEAVE_ZONE);
+			isActive = false;
+			return;
+		}
+	}
+	isActive = true;
+	timeElapsed += 1;
```

Comment:
```suggestion
	++timeElapsed;
```

Reply:
I am curious about using the prefixed version—is there a specific performance or logic reason for doing it that way, rather than `timeElapsed++`? Or is it simply a style choice?

Reply:
For performance reasons in the general case yes, for example with iterators it matters. For ints it doesn't matter, but for consistency preincrement is also used there.